File: app/src/main/java/cn/xhuww/mvvm/widget/RippleTextView.java
package cn.xhuww.mvvm.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.support.v7.widget.AppCompatTextView;

/**
 * Created by xhu_ww on 2018/5/9.
 * description:水波纹效果 TextView
 */
public class RippleTextView extends AppCompatTextView {
    //画笔
    private Paint mPaint;
    //波纹的中心点
    private PointF centPoint;
    //波纹初始半径
    private int initRadius;
    private float maxRadius;
    //用于判断是否接收up事件
    private boolean isPressed;
    private boolean isShowRipple;

    public RippleTextView(Context context) {
        this(context, null);
    }

    public RippleTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RippleTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        isShowRipple = Build.VERSION.SDK_INT >= 18;
    }

    private void setCentPoint(float x, float y) {
        if (isShowRipple) {
            this.centPoint = new PointF(x, y);
            if (mPaint == null) {
                mPaint = new Paint();
                mPaint.setAntiAlias(true);
                mPaint.setAlpha(40);
                mPaint.setStyle(Paint.Style.FILL);
            }
            initRadius = 0;
            maxRadius = x >= getWidth() / 2 ? x : getWidth() - x;
            invalidate();
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        if (isShowRipple) {
            if (event.getAction() == MotionEvent.ACTION_UP && !isPressed) {
                isPressed = true;
                setCentPoint(event.getX(), event.getY());
                //先让up事件吊起，当波纹动作完成后再传up事件
                postDelayed(new DispatchUpTouchEventRunnable(event), (long) (maxRadius / 20) * 10 + 400);
                return true;
            } else if (event.getAction() == MotionEvent.ACTION_DOWN) {
                isPressed = false;
            }
        }
        return true;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (centPoint != null && isShowRipple) {
            if ((initRadius += 20) <= maxRadius) {
                canvas.drawCircle(centPoint.x, centPoint.y, initRadius, mPaint);
            } else {
                centPoint = null;
            }
            postInvalidateDelayed(10);
        }
    }

    class DispatchUpTouchEventRunnable implements Runnable {
        public MotionEvent event;

        public DispatchUpTouchEventRunnable(MotionEvent event) {
            this.event = event;
        }

        @Override
        public void run() {
            RippleTextView.this.dispatchTouchEvent(event);
        }
    }
}

File: kotlin_mvvm/src/main/java/com/example/study/widget/RippleTextView.kt
package com.example.study.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import android.os.Build
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet
import android.view.MotionEvent

/**
 * Created by xhu_ww on 2018/6/8.
 * description: 水波纹TextView
 */
class RippleTextView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {

    //画笔
    private var mPaint: Paint? = null
    //波纹的中心点
    private var centPoint: PointF? = null
    //波纹初始半径
    private var initRadius = 0
    private var maxRadius = 0f

    private var isPress = false
    private val isShowRipple = Build.VERSION.SDK_INT >= 18

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (centPoint != null && isShowRipple) {
            initRadius += 20
            if (initRadius <= maxRadius) {
                canvas.drawCircle(centPoint!!.x, centPoint!!.y, initRadius.toFloat(), mPaint)
            } else {
                centPoint = null
            }
            postInvalidateDelayed(10)
        }
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        if (isShowRipple) {
            if (event.action == MotionEvent.ACTION_UP && !isPress) {
                isPress = true
                setCentPoint(event.x, event.y)
                //先让up事件吊起，当波纹动作完成后再传up事件
                postDelayed(DispatchUpTouchEventRunnable(event), (maxRadius / 20).toLong() * 10 + 400)
            } else if (event.action == MotionEvent.ACTION_DOWN) {
                isPress = false
            }
        }
        return true
    }

    private fun setCentPoint(x: Float, y: Float) {
        if (isShowRipple) {
            this.centPoint = PointF(x, y)
            if (mPaint == null) {
                mPaint = Paint()
                mPaint!!.isAntiAlias = true
                mPaint!!.alpha = 40
                mPaint!!.style = Paint.Style.FILL
            }
            initRadius = 0
            maxRadius = if (x >= width / 2) x else width - x
            invalidate()
        }
    }

    internal inner class DispatchUpTouchEventRunnable(var event: MotionEvent) : Runnable {

        override fun run() {
            this@RippleTextView.dispatchTouchEvent(event)
        }
    }
}
